pkg/derivation/derivation.go

```diff
@@ -81,29 +86,163 @@ func (d *Derivation) Validate() error {
 	return nil
 }
 
+func compressHash(hash []byte, newSize int) []byte {
+	buf := make([]byte, newSize)
+	for i := 0; i < len(hash); i++ {
+		buf[i%newSize] ^= hash[i]
+	}
+	return buf
+}
+
+func (d *Derivation) Name() string {
+	for _, e := range d.EnvVars {
+		if e.Key == "name" {
+			return e.Value
+		}
+	}
+
+	// TODO: Maybe panic or change type sig to (string, error)?
+	return ""
+}
+
+func (d *Derivation) OutputPaths(store KVStore) (map[string]string, error) {
+
+	_, ok := store.(*InputSubstKV)
+	if !ok {
+		store = NewInputSubstKV(store)
+	}
+
+	name := d.Name()
+
+	var buf bytes.Buffer
+	err := d.writeDerivation(&buf, true, store)
+	if err != nil {
+		return nil, err
+	}
+	b := buf.Bytes()
+
+	h := sha256.New()
+	_, err = h.Write(b)
+	if err != nil {
+		return nil, err
+	}
+	partialDrvHash := hex.EncodeToString(h.Sum(nil))
+
+	outputs := make(map[string]string)
+	for _, o := range d.Outputs {
+		fixed := d.FixedOutput()
+		if fixed != nil {
+			s := fmt.Sprintf("source:sha256:%s:%s:%s", o.Hash, nixpath.StoreDir, name)
+
+			h := sha256.New()
+			_, err = h.Write([]byte(s))
+			if err != nil {
+				return nil, err
+			}
+			digest := h.Sum(nil)
+
+			outputs[o.Name] = filepath.Join(nixpath.StoreDir, nixbase32.EncodeToString(compressHash(digest, 20))+"-"+name)
+			continue
+		}
+
+		outputSuffix := name
+		if o.Name != "out" {
+			outputSuffix += "-" + o.Name
+		}
+
+		s := fmt.Sprintf("output:%s:sha256:%s:%s:%s", o.Name, partialDrvHash, nixpath.StoreDir, outputSuffix)
+
+		h := sha256.New()
+		_, err = h.Write([]byte(s))
+		if err != nil {
+			return nil, err
+		}
+		digest := h.Sum(nil)
+
+		outputs[o.Name] = filepath.Join(nixpath.StoreDir, nixbase32.EncodeToString(compressHash(digest, 20))+"-"+outputSuffix)
+	}
+
+	return outputs, nil
+}
+
+// FixedOutput - Returns the fixed output any is found, otherwise returns nil.
```

Comment:
Should this be public api? What are the consumers?

What happens if there's more than one output where `HashAlgorithm` is the empty string?

Reply:
I think this is related to 95f413a6e5edc8b88e56e144dc2e83676d8551a7.